algorithms/src/snark/marlin/ahp/ahp.rs

```diff
@@ -142,114 +142,161 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+            }).collect::<Vec<F>>();
+            z_b_i_s
+        }).collect_vec();
+        let batch_z_b_at_beta: F = z_b_s_at_beta.iter()
+                                                .zip_eq(batch_combiners.values())
+                                                .zip_eq(r_alpha_at_beta_s.iter())
+                                                .map(|((z_b_i_at_beta, combiners), &r_alpha_at_beta)| {
+            let z_b_at_beta = z_b_i_at_beta.iter()
+                                           .zip_eq(combiners.instance_combiners.iter())
+                                           .map(|(z_b_at_beta, instance_combiner)| {
+                *z_b_at_beta * instance_combiner
+            }).sum::<F>();
+            r_alpha_at_beta * z_b_at_beta
+        }).sum();
         let g_1_at_beta = evals.get_lc_eval(&g_1, beta)?;
 
-        let lag_at_beta = input_domain.evaluate_all_lagrange_coefficients(beta);
-        let combined_x_at_beta = batch_combiners
-            .iter()
-            .zip_eq(&public_inputs)
-            .map(|(c, x)| x.iter().zip_eq(&lag_at_beta).map(|(x, l)| *x * l).sum::<F>() * c)
-            .sum::<F>();
+        let mut combined_x_at_betas: Vec<F> = Vec::with_capacity(state.circuit_specific_states.len());
+        for (i, (circuit_id, circuit_state)) in state.circuit_specific_states.iter().enumerate() {
+            let lag_at_beta = circuit_state.input_domain.evaluate_all_lagrange_coefficients(beta);
+            let combined_x_at_beta = batch_combiners[circuit_id]
+                .instance_combiners
+                .iter()
+                .zip_eq(parsed_public_inputs[i].iter())
+                .map(|(c, x)| x.iter().zip_eq(&lag_at_beta).map(|(x, l)| *x * l).sum::<F>() * c)
+                .sum::<F>();
+            combined_x_at_betas.push(combined_x_at_beta);
+        }
+
+        // TODO: where to multiply the selectors and circuit_combiners? To elements here or just before creating the (virtual) commitment?
+        // let batch_combiners[].circuit_combiner
+        // let selector_i = Self::get_selector_evaluation(&cached_selectors, &largest_constraint_domain, &circuit_state.constraint_domain, beta);
 
         #[rustfmt::skip]
         let lincheck_sumcheck = {
             let mut lincheck_sumcheck = LinearCombination::empty("lincheck_sumcheck");
             if MM::ZK {
                 lincheck_sumcheck.add(F::one(), "mask_poly");
             }
-            for (i, (z_b_i_at_beta, combiner)) in z_b_s_at_beta.iter().zip_eq(batch_combiners).enumerate() {
+            for (i, (z_b_i_at_beta, (circuit_id, combiners))) in z_b_s_at_beta.iter().zip_eq(batch_combiners).enumerate() {
+                for (j, ((z_b_i_j_at_beta, instance_combiner), &r_alpha_at_beta)) in z_b_i_at_beta.iter()
+                                                            .zip_eq(combiners.instance_combiners.iter())
+                                                            .zip_eq(r_alpha_at_beta_s.iter())
```

Comment:
cannot move out of `combiners.instance_combiners` which is behind a shared reference
move occurs because `combiners.instance_combiners` has type `snarkvm_utilities::Vec<F>`, which does not implement the `Copy` trait

Reply:
Done